pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/LinuxInfoUtils.java

```diff
@@ -85,6 +86,8 @@ public static double getTotalCpuLimit(boolean isCGroupsEnabled) {
                 if (quota > 0) {
                     return 100.0 * quota / period;
                 }
+                int cpuCount = readTrimStringFromFile(Paths.get(CGROUPS_CPU_USAGE_PER_CPU_PATH)).split(" ").length;
```

Comment:
Could you provide some references regarding this change?

Though this appears to be a good way to estimate the number of cpus, I wonder if this is the "correct" way to get the max number of CPUs per cgroup.

Reply:
```
- cpuacct.usage_percpu
reports the CPU time (in nanoseconds) consumed on each CPU by all tasks in this cgroup (including tasks lower in the hierarchy).
- cpuacct.usage
reports the total CPU time (in nanoseconds) consumed by all tasks in this cgroup (including tasks lower in the hierarchy).
```
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/resource_management_guide/sec-cpuacct

The sum of the values in cpuacct.usage_percpu is equal to the value in cpuacct.usage,and we have calculated the cpu usage base on cpuacct.usage. Using cpuacct.usage_percpu to calculate cpu limit can ensure that the highest cpu usage is 100%.

Reply:
I see the issue is that,  if `cfs_quota_us` is -1, then we set the cpu limit based on the JVM, but still get the cpu usage from the cgroup usage sys files.

```
private double getTotalCpuUsage(double elapsedTimeSeconds) {
        if (isCGroupsEnabled) {
            return getTotalCpuUsageForCGroup(elapsedTimeSeconds);
        } else {
            return getTotalCpuUsageForEntireHost();
        }
    }
```



So please put the error message when setting the jvm limit. Ideally, I think we should never use jvm limit or get the cpu usage from the jvm to match.
```
// Fallback to JVM reported CPU quota
        return 100 * Runtime.getRuntime().availableProcessors();
```

From my understanding, `cfs_quota_us` is -1 means there are no CPU time restrictions, which means all "effective" cpus can be used. 

but I am worried that if the cgroup happens to hit smaller sets cpus at that time, then `cpuCount` here could be less than what the cgroup actually configured(effective cpus).


should we rely on cpuset (effective cpus) instead for  `cpuCount` ?
`cpuset.cpus` or `cpuset.cpus.effective`

Reply:
I am not expert in CGROUP, but it seems that cpuset.cpus would be a good choice. cpuset.cpus is a mandatory parameter in cpuset subsystem, and cpuset.cpus.effective is not mandatory.
for example, i do not find cpuset.cpus.effective in my system,but cpuset.effective_cpus.

```
cpuset.cpus (mandatory)
specifies the CPUs that tasks in this cgroup are permitted to access. This is a comma-separated list, with dashes ("-") to represent ranges. For example,
0-2,16
represents CPUs 0, 1, 2, and 16.
```

Reply:
i have changed the implementation.